namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala

```diff
@@ -82,12 +82,12 @@ private[consul] object SvcAddr {
             )
             stopped = true
             Future.Unit
-            // this exception case checks if we queried for a service in a datacenter that
-            // doesn't exist. We capture this case so that we can return Addr.Neg to prevent
-            // service name resolution from timing out. Since Consul's HTTP 5xx API response is
-            // overloaded and represents different error states, its necessary for us to capture this
-            // exception and resolve to Addr.Neg see https://github.com/hashicorp/consul/issues/4901
-          case Throw(e: UnexpectedResponse) if e.rsp.contentString == DatacenterErrorMessage =>
+          // this exception case checks if we queried for a service in a datacenter that
+          // doesn't exist. We capture this case so that we can return Addr.Neg to prevent
+          // service name resolution from timing out. Since Consul's HTTP 5xx API response is
+          // overloaded and represents different error states, its necessary for us to capture this
+          // exception and resolve to Addr.Neg see https://github.com/hashicorp/consul/issues/4901
+          case Throw(e: UnexpectedResponse) if e.rsp.contentString == DatacenterErrorMessage && currentState == Addr.Pending =>
```

Comment:
Now I think, do we really need to parse the error message? Don't we want to have the same behavior as we have on this line for all errors? Wouldn't it be correct to return `Neg`on any error if we haven't resolved anything yet instead of just stucking in `Pending`?

This would simplify code as well.

Reply:
I think you're right.  The difference in behavior from today would be that if we get a 5XX which is not a "No path to DC" error while the Addr is pending, we would immediately update it to Neg (and then retry) instead of leaving it as Pending and retrying.  But I think that's okay.

It's been a while since I've read this code so I'm trying to remember how it all works.  I don't see anything in here for dealing with 404s.  Do you remember how SvcAddr deals with service deletion?  When a service is deleted, does that catalog call just return an empty address list?

Reply:
> The difference in behavior from today would be that if we get a 5XX which is not a "No path to DC" error while the Addr is pending, we would immediately update it to Neg (and then retry) instead of leaving it as Pending and retrying. But I think that's okay.

Pending is generally bad, because, I believe, it will make linkerd to timeout eventually without letting namerd exploring other branches of dtab.

Let's describe all possible kinds of errors we can expect from consul
1. Consul agent can't be contacted
already handled in code partially by `Throw(e: IndividualRequestTimeoutException)`, partially by `Throw(e)`
Namer should retain last known state :heavy_check_mark:
If it is an initial request, Namer should return `Addr.Neg` :question: Currently it is not the case, and I think it is not entirely correct. `Addr.Pending` will be returned right now and if it remains like that for more than 10 seconds, linkerd will return 502. If it was `Addr.Neg`, namerd could try other discovery systems in the meantime (for example, try dns, or try contacting consul server directly instead of local agent (given proper namerd configuration and dtab of course)). 
Basically it is a choice between "let requests hang for 10 seconds and then fail" and "let requests fail right away". But with the latter option we can try other discovery systems before failing (calls to other consul datacenters will likely fail too).

2. Consul agent can be contacted but its response is not a response we expect
already handled in code by `Throw(e)`
Basically same as case 1. In the code the only difference is the message logged.

3. Consul agent can be contacted but, but not all servers can be contacted (No path to datacenter error)
Handled in code by `Throw(e: UnexpectedResponse) if e.rsp.contentString == DatacenterErrorMessage`
Namerd should retain last known state :heavy_check_mark:  (fixed with this commit)
If it is an initial request, Namer should return `Addr.Neg`.
Here, again, it is a choice between "let requests hang for 10 seconds and then fail" and "let requests fail right away". However, with the second option we can try other datacenters as well.

Would you agree, that behavior in those 3 cases looks very similar, and in code it is actually the same, so 3 cases can be merged into one? Only logged messages should be different.

Reply:
@edio This is a great write-up!  Yes, I agree, I think it makes sense to merge these 3 cases.

Reply:
@adleong I'll fix the code shortly. Thanks!

Reply:
@dadjeibaah , my point is that other cases should behave the same way "No path to datacenter" case behaves after this PR.

If a name was never resolved — return `Addr.Neg` and let fallback to other parts of dtab
If a name was ever resolved before — return result of previous resolution

The code that handles all cases will look like this (still working on tests, so can't push)

```
          case Throw(e) =>
            // Update state only if it is Pending to not let linkerd hang on name resolution
            // Otherwise retain last known state to allow namerd survive intermittent failures
            val effectiveState = if (currentState == Addr.Pending) {
              state.update(Addr.Neg)
              Addr.Neg
            } else {
              currentState
            }

            log.log(
              failureLogLevel,
              "consul datacenter %s service %s lookup request received invalid dc error %s. Current state is %s",
              datacenter, key.name, e, effectiveState
            )
            val backoff #:: nextBackoffs = backoffs
            // subsequent errors are logged as DEBUG
            Future.sleep(backoff).before(loop(None, nextBackoffs, Level.DEBUG, effectiveState))
```

Let me know please if I'm still missing something with this chane.